metadata/src/utils/retain.rs

```diff
@@ -4,49 +4,278 @@
 
 //! Utility functions to generate a subset of the metadata.
 
+use alloc::collections::BTreeSet;
+
 use crate::{
     ExtrinsicMetadata, Metadata, OuterEnumsMetadata, PalletMetadataInner, RuntimeApiMetadataInner,
     StorageEntryType,
 };
-use alloc::collections::BTreeMap;
-use hashbrown::HashSet;
-use scale_info::TypeDef;
-
-/// Collect all type IDs needed to represent the provided pallet.
-fn collect_pallet_types(pallet: &PalletMetadataInner, type_ids: &mut HashSet<u32>) {
-    if let Some(storage) = &pallet.storage {
-        for entry in storage.entries() {
-            match entry.entry_type {
-                StorageEntryType::Plain(ty) => {
-                    type_ids.insert(ty);
+use alloc::collections::{BTreeMap, VecDeque};
+use alloc::vec::Vec;
+use scale_info::{
+    PortableType, TypeDef, TypeDefArray, TypeDefBitSequence, TypeDefCompact, TypeDefComposite,
+    TypeDefSequence, TypeDefTuple, TypeDefVariant,
+};
+
+use super::variant_index::VariantIndex;
+
+#[derive(Clone)]
+struct TypeSet {
+    seen_ids: BTreeSet<u32>,
+}
+
+impl TypeSet {
+    fn new() -> Self {
+        Self {
+            seen_ids: BTreeSet::new(),
+        }
+    }
+
+    fn collect_types(&mut self, metadata: &Metadata, t: &PortableType) {
```

Comment:
So this logic is basically recursing through all of the types linked from some type and adding them to the set. But this is the same thing that we do in `PortableRegistry::retain` here https://github.com/paritytech/scale-info/blob/master/src/portable.rs#L104, so I don't really understand why it's necessary and not just duplicating the same work offhand?

Reply:
Yup, retain is called dead last before updating types, but we need to collect types ourselves to determine which outer type variants are safe to drop/which pallets can be stripped or dropped completetely.

Reply:
I'm not sure how I feel about this function, because it duplicates the logic in scale_info's `retain`.

a) On the one hand, we could also remove this function entirely (ie no longer recurse and building this type Ids map when we collect up our pallet/runtime etc types), and instead have a function which has the specific goal of searching through the metadata to look for the runtime types (and in that way, it doesn't need to allocate/collect anything).

b) On the other hand, keeping this function is "easier" because you need to recurse anyway to do the above searching, so maybe you end up with fewer lines of code by just keeping it around.

Personally I'd try (a) and see whether doing so makes things easier to follow or not, but not sure!